semantic_model_generator/data_processing/cte_utils.py

```diff
@@ -125,6 +145,79 @@ def _generate_cte_for(
         return cte
 
 
+def get_all_physical_column_references(
+    column: semantic_model_pb2.Column,
+) -> List[str]:
+    """Returns a set of column names referenced in the column expression.
+
+    For example, the following column expressions yield the following return values:
+    foo -> [foo]
+    foo+bar -> [foo, bar]
+    sum(foo) -> [foo]
+    """
+    try:
+        parsed = sqlglot.parse_one(column.expr, dialect=Snowflake)
+        col_names = set()
+        for col in parsed.find_all(sqlglot.expressions.Column):
+            # TODO(renee): Handle quoted columns.
+            col_name = col.name.lower()
+            if col.this.quoted:
+                col_name = col.name
+            col_names.add(col_name)
+        return sorted(list(col_names))
+    except Exception as ex:
+        raise ValueError(f"Failed to parse sql expression: {column.expr}. Error: {ex}")
+
+
+def direct_mapping_logical_columns(
+    table: semantic_model_pb2.Table,
+) -> List[semantic_model_pb2.Column]:
+    """
+    Returns a list of logical columns that map 1:1 to an underlying physical column
+    (i.e. logical table's expression is simply the physical column name) in this table.
+    """
+    ret: List[semantic_model_pb2.Column] = []
+    for c in table.columns:
+        if _is_physical_table_column(c):
+            ret.append(c)
+    return ret
+
+
+def _enrich_column_in_expr_with_aggregation(
+    table: semantic_model_pb2.Table,
+) -> semantic_model_pb2.Table:
+    """
+    Expands the logical columns of 'table' to include columns mentioned in a logical columns
+    with an aggregate expression. E.g. for a logical column called CPC with expr sum(cost) / sum(clicks),
+    adds logical columns for "cost" and "clicks", in not present.
+    """
+    direct_mapping_lcols = [
+        c.name.lower() for c in direct_mapping_logical_columns(table)
+    ]
+    cols_to_append = set()
+    for col in table.columns:
+        if not is_aggregation_expr(col):
+            continue
+        for pcol in get_all_physical_column_references(col):
+            # If the physical column doesn't have a direct mapping logical column
+            # with the same name, then we need to add a new logical column for it.
```

Comment:
add comment: this may introduce duplicated physical table added; we'll improve later?